src/block_storage/volumes.rs

```diff
@@ -0,0 +1,453 @@
+// Copyright 2024 Sandro-Alessio Gierens <sandro@gierens.de>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Volume management via Block Storage API.
+
+use async_trait::async_trait;
+use futures::stream::{Stream, TryStreamExt};
+use std::collections::HashMap;
+use std::fmt::{self, Display, Formatter};
+use std::time::Duration;
+
+use super::super::common::{Refresh, ResourceIterator, ResourceQuery};
+use super::super::session::Session;
+use super::super::utils::Query;
+use super::super::waiter::DeletionWaiter;
+use super::super::{Result, Sort};
+use super::{api, protocol};
+
+/// A query to volume list.
+#[derive(Clone, Debug)]
+pub struct VolumeQuery {
+    session: Session,
+    query: Query,
+    can_paginate: bool,
+    sort: Vec<String>,
+}
+
+/// Structure representing a summary of a single volume.
+#[derive(Clone, Debug)]
+pub struct Volume {
+    session: Session,
+    inner: protocol::Volume,
+}
+
+/// A request to create a volume.
+#[derive(Clone, Debug)]
+pub struct NewVolume {
+    session: Session,
+    inner: protocol::VolumeCreate,
+}
+
+impl Display for Volume {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:#?}", self.inner)
+    }
+}
+
+impl Volume {
+    /// Create an Volume object.
+    pub(crate) async fn new<Id: AsRef<str>>(session: Session, id: Id) -> Result<Volume> {
+        let inner = api::get_volume(&session, id).await?;
+        Ok(Volume { session, inner })
+    }
+
+    transparent_property! {
+        #[doc = "Migration status."]
+        migration_status: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Volume attachments."]
+        attachments: ref Vec<protocol::VolumeAttachment>
+    }
+
+    transparent_property! {
+        #[doc = "Volume links."]
+        links: ref Vec<protocol::Link>
+    }
+
+    transparent_property! {
+        #[doc = "Name of the availability zone."]
+        availability_zone: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Current backend of the volume."]
+        host: ref Option<String>
+    }
+
+    transparent_property! {
+        #[doc = "Whether the volume is encrypted."]
+        encrypted: ref bool
```

Comment:
You don't need to use ref with fields whose type is Copy, this leads to rather unergonomic `&bool` instead of just `bool`. The `ref` modifier is only needed for fields that cannot be just copied.

Reply:
done